### PR TITLE
Add Epic Issues API

### PR DESCRIPTION
**Pull request by justinrush**
This PR adds the methods from the [Epic Issues API](https://docs.gitlab.com/ee/api/epic_issues.html):
* ListEpicIssues
* AssignEpicIssue
* RemoveEpicIssue
* UpdateEpicIssueAssignment

### CODE REVIEW

File: epic_issues.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+// EpicIssueAssignment contains both the Epic and Issue objects returned from Gitlab w/ the assignment ID
+type EpicIssueAssignment struct {
+	ID    int    `json:"id"`
+	Epic  *Epic  `json:"epic"`
+	Issue *Issue `json:"issue"`
+}
+
+// AssignEpicIssue assigns an existing issue to an Epic
```
Comment (svanharmelen) [inline epic_issues.go:38]: In Go comments should be treated as sentences: https://blog.golang.org/godoc

```suggestion
// AssignEpicIssue assigns an existing issue to an Epic.
```

File: epic_issues.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+//
+// Gitlab API Docs: https://docs.gitlab.com/ee/api/epic_issues.html#assign-an-issue-to-the-epic
+func (s *EpicsService) AssignEpicIssue(gid interface{}, epic int, issue int, options ...RequestOptionFunc) (*EpicIssueAssignment, *Response, error) {
+	group, err := parseID(gid)
+	if err != nil {
+		return nil, nil, err
+	}
+
```
Comment (svanharmelen) [inline epic_issues.go:46]:
```suggestion
```

File: epic_issues.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	u := fmt.Sprintf("groups/%s/epics/%d/issues/%d", pathEscape(group), epic, issue)
+
+	req, err := s.client.NewRequest("POST", u, nil, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var a *EpicIssueAssignment
+
```
Comment (svanharmelen) [inline epic_issues.go:55]:
```suggestion
```

File: epic_issues.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	resp, err := s.client.Do(req, &a)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return a, resp, err
+}
+
+// RemoveEpicIssue removes an issue from an Epic
```
Comment (svanharmelen) [inline epic_issues.go:64]:
```suggestion
// RemoveEpicIssue removes an issue from an Epic.
```

File: epic_issues.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+//
+// Gitlab API Docs: https://docs.gitlab.com/ee/api/epic_issues.html#remove-an-issue-from-the-epic
```
Comment (svanharmelen) [inline epic_issues.go:66]:
```suggestion
// Gitlab API Docs:
// https://docs.gitlab.com/ee/api/epic_issues.html#remove-an-issue-from-the-epic
```

File: epic_issues.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+func (s *EpicsService) RemoveEpicIssue(gid interface{}, epic int, epicIssue int, options ...RequestOptionFunc) (*EpicIssueAssignment, *Response, error) {
+	group, err := parseID(gid)
+	if err != nil {
+		return nil, nil, err
+	}
+
```
Comment (svanharmelen) [inline epic_issues.go:72]:
```suggestion
```

File: epic_issues.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	u := fmt.Sprintf("groups/%s/epics/%d/issues/%d", pathEscape(group), epic, epicIssue)
+
+	req, err := s.client.NewRequest("DELETE", u, nil, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var a *EpicIssueAssignment
+
```
Comment (svanharmelen) [inline epic_issues.go:81]:
```suggestion
```

File: epic_issues.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	resp, err := s.client.Do(req, &a)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return a, resp, err
+}
+
+// UpdateEpicIsssueAssignmentOptions describes options to move issues within an epic
```
Comment (svanharmelen) [inline epic_issues.go:90]: Same thing about the comment and doc reference.

File: epic_issues.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+type UpdateEpicIsssueAssignmentOptions struct {
+	*ListOptions
+	MoveBeforeID int `json:"move_before_id"`
+	MoveAfterID  int `json:"move_after_id"`
```
Comment (svanharmelen) [inline epic_issues.go:94]: The last two fields should both be `*int` and the tags need to include tags for `url` and both need to contain `omitemtpy`.

File: epic_issues.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+// UpdateEpicIssueAssignment moves an issue before or after another issue in an epic issue list
```
Comment (svanharmelen) [inline epic_issues.go:97]: Same about the comment...

File: epic_issues.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+//
+// Gitlab API Docs: https://docs.gitlab.com/ee/api/epic_issues.html#update-epic---issue-association
```
Comment (svanharmelen) [inline epic_issues.go:99]:
```suggestion
// Gitlab API Docs:
// https://docs.gitlab.com/ee/api/epic_issues.html#update-epic---issue-association
```

File: epic_issues.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+func (s *EpicsService) UpdateEpicIssueAssignment(gid interface{}, epic int, epicIssue int, opt *UpdateEpicIsssueAssignmentOptions, options ...RequestOptionFunc) ([]*Issue, *Response, error) {
+	group, err := parseID(gid)
+	if err != nil {
+		return nil, nil, err
+	}
+
```
Comment (svanharmelen) [inline epic_issues.go:105]:
```suggestion
```

File: epic_issues.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+type EpicIssueAssignment struct {
+	ID    int    `json:"id"`
+	Epic  *Epic  `json:"epic"`
+	Issue *Issue `json:"issue"`
+}
+
+// AssignEpicIssue assigns an existing issue to an Epic
+//
+// Gitlab API Docs: https://docs.gitlab.com/ee/api/epic_issues.html#assign-an-issue-to-the-epic
+func (s *EpicsService) AssignEpicIssue(gid interface{}, epic int, issue int, options ...RequestOptionFunc) (*EpicIssueAssignment, *Response, error) {
```
Comment (svanharmelen) [inline epic_issues.go:41]: When declaring two params of the same type, you can write that a bit nicer/shorter:
```suggestion
func (s *EpicsService) AssignEpicIssue(gid interface{}, epic, issue int, options ...RequestOptionFunc) (*EpicIssueAssignment, *Response, error) {
```

File: epic_issues.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+	u := fmt.Sprintf("groups/%s/epics/%d/issues/%d", pathEscape(group), epic, epicIssue)
+
+	req, err := s.client.NewRequest("PUT", u, opt, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var i []*Issue
+
```
Comment (svanharmelen) [inline epic_issues.go:114]:
```suggestion
```

File: epic_issues.go [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	u := fmt.Sprintf("groups/%s/epics/%d/issues/%d", pathEscape(group), epic, epicIssue)
+
+	req, err := s.client.NewRequest("PUT", u, opt, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var i []*Issue
```
Comment (svanharmelen) [inline epic_issues.go:113]: Same thing about the name (maybe add an `s`).

File: issues.go [hunk 16]
```diff
@@ -111,6 +111,7 @@ type Issue struct {
 	Links                *IssueLinks      `json:"_links"`
 	IssueLinkID          int              `json:"issue_link_id"`
 	MergeRequestCount    int              `json:"merge_requests_count"`
+	EpicIssueID          int              `json:"epic_issue_id"`
```
[the remark this answers is not in the document]
Comment (justinrush) [reply issues.go:114]: I wasn't really sure how to handle this.  apparently when issues are returned via the ListEpicIssues API it returns an array of issues that have the epic_issue_id attribute on them.  That id is needed to remove or update the association.

I thought about just creating a new EpicIssue struct and inheriting the Issue struct, but there didn't really seem to be a downside including it in the main Issue struct and we don't introduce another object type.  I can go either way - what's your preference?

[the remark this answers is not in the document]
Comment (svanharmelen) [reply issues.go:114]: With this context I’m fine with leaving this as is 👍🏻

**Comment by justinrush** [comment]
@svanharmelen thanks for patiently dealing with my style issues - still new to golang and i rely on the linters pretty heavily and sometimes they stop working on me.  I think I addressed all your feedback - let me know if there is anything else.

**Pull request merged**